app/services/solidus_square/payment_sync_service.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+module SolidusSquare
+  class PaymentSyncService < Base
+    attr_reader :params
+
+    def initialize(params)
+      @params = params
+
+      super()
+    end
+
+    def call
+      create_square_payment! if hosted_checkout?
+
+      update_payment_source!
+      payment.complete if payment_source_captured?
+
+      complete_order!
+    end
+
+    private
+
+    def complete_order!
+      return if order.complete?
```

Comment:
Considering we call `next!` and `complete!` it means we assume that the order should be in a payment state here, so I think we can replace that code with an exit condition when the order in not in payment state.

Reply:
https://github.com/nebulab/solidus_square/pull/43/commits/abea86ce4b352a0e46b124beeecf97797b17300c